File: examples/python/bitwarden_client.py
"""
bitwarden_client.py — Retrieve secrets from Bitwarden via the bw CLI.

Prerequisites:
    pip install -r requirements.txt
    bw login (or bw login --apikey)
    export BW_SESSION=$(bw unlock --raw)

Usage:
    python bitwarden_client.py
"""

import json
import os
import subprocess
import sys


class BitwardenClient:
    """Thin wrapper around the Bitwarden CLI (`bw`)."""

    def __init__(self, session: str | None = None):
        self.session = session or os.environ.get("BW_SESSION", "")
        if not self.session:
            raise EnvironmentError(
                "BW_SESSION is not set. Run: export BW_SESSION=$(bw unlock --raw)"
            )
        self._verify_unlocked()

    def _run(self, args: list[str], check: bool = True) -> subprocess.CompletedProcess:
        """Run a bw CLI command with the session key."""
        cmd = ["bw"] + args + ["--session", self.session, "--nointeraction"]
        return subprocess.run(cmd, capture_output=True, text=True, check=check)

    def _verify_unlocked(self) -> None:
        result = self._run(["status"], check=False)
        if result.returncode != 0 or '"status":"unlocked"' not in result.stdout:
            raise PermissionError("Bitwarden vault is locked or session key is invalid.")

    def sync(self) -> None:
        """Sync vault from server."""
        self._run(["sync"])

    def get_password(self, item_name: str) -> str:
        """Retrieve the password for a login item by name."""
        result = self._run(["get", "password", item_name], check=False)
        if result.returncode != 0 or not result.stdout.strip():
            raise KeyError(f"Item '{item_name}' not found or has no password.")
        return result.stdout.strip()

    def get_username(self, item_name: str) -> str:
        """Retrieve the username for a login item by name."""
        result = self._run(["get", "username", item_name])
        return result.stdout.strip()

    def get_item(self, item_name: str) -> dict:
        """Retrieve the full item JSON by name."""
        result = self._run(["get", "item", item_name], check=False)
        if result.returncode != 0:
            raise KeyError(f"Item '{item_name}' not found.")
        return json.loads(result.stdout)

    def get_field(self, item_name: str, field_name: str) -> str:
        """Retrieve a custom field value from an item."""
        item = self.get_item(item_name)
        fields = item.get("fields") or []
        for field in fields:
            if field.get("name") == field_name:
                return field.get("value", "")
        raise KeyError(f"Field '{field_name}' not found in item '{item_name}'.")

    def get_note(self, item_name: str) -> str:
        """Retrieve the notes of an item (e.g. a certificate or config blob)."""
        result = self._run(["get", "notes", item_name])
        return result.stdout.strip()

    def list_items(self, search: str | None = None) -> list[dict]:
        """List vault items, optionally filtered by a search term."""
        args = ["list", "items"]
        if search:
            args += ["--search", search]
        result = self._run(args)
        return json.loads(result.stdout)


def demo() -> None:
    try:
        client = BitwardenClient()
    except (EnvironmentError, PermissionError) as e:
        print(f"ERROR: {e}", file=sys.stderr)
        sys.exit(1)

    # Search for items
    items = client.list_items(search="github")
    print(f"Found {len(items)} item(s) matching 'github'")

    if items:
        item = items[0]
        print(f"First match: {item.get('name')}")

        # Get password (don't print in real code)
        try:
            password = client.get_password(item["name"])
            print(f"Password retrieved (length: {len(password)})")
        except KeyError as e:
            print(f"Could not retrieve password: {e}")


if __name__ == "__main__":
    demo()

File: examples/go/bitwarden_client.go
// bitwarden_client.go — Retrieve secrets from Bitwarden via the bw CLI.
//
// Prerequisites:
//   go mod tidy
//   bw login (or bw login --apikey)
//   export BW_SESSION=$(bw unlock --raw)
//
// Usage:
//   go run bitwarden_client.go

package main

import (
	"encoding/json"
	"fmt"
	"log"
	"os"
	"os/exec"
	"strings"
)

// BitwardenItem represents a Bitwarden vault item (partial).
type BitwardenItem struct {
	ID    string `json:"id"`
	Name  string `json:"name"`
	Notes string `json:"notes"`
	Login struct {
		Username string `json:"username"`
		Password string `json:"password"`
	} `json:"login"`
	Fields []struct {
		Name  string `json:"name"`
		Value string `json:"value"`
	} `json:"fields"`
}

// BitwardenClient wraps the bw CLI.
type BitwardenClient struct {
	session string
}

// NewBitwardenClient creates a client using the BW_SESSION environment variable.
func NewBitwardenClient() (*BitwardenClient, error) {
	session := os.Getenv("BW_SESSION")
	if session == "" {
		return nil, fmt.Errorf("BW_SESSION is not set — run: export BW_SESSION=$(bw unlock --raw)")
	}
	c := &BitwardenClient{session: session}
	if err := c.verifyUnlocked(); err != nil {
		return nil, err
	}
	return c, nil
}

func (c *BitwardenClient) run(args ...string) (string, error) {
	args = append(args, "--session", c.session, "--nointeraction")
	cmd := exec.Command("bw", args...)
	out, err := cmd.Output()
	if err != nil {
		return "", fmt.Errorf("bw %s: %w", strings.Join(args[:len(args)-2], " "), err)
	}
	return strings.TrimSpace(string(out)), nil
}

func (c *BitwardenClient) verifyUnlocked() error {
	out, err := c.run("status")
	if err != nil || !strings.Contains(out, `"status":"unlocked"`) {
		return fmt.Errorf("Bitwarden vault is locked or session key is invalid")
	}
	return nil
}

// Sync syncs the vault from the server.
func (c *BitwardenClient) Sync() error {
	_, err := c.run("sync")
	return err
}

// GetItem retrieves a full item by name.
func (c *BitwardenClient) GetItem(itemName string) (*BitwardenItem, error) {
	out, err := c.run("get", "item", itemName)
	if err != nil {
		return nil, fmt.Errorf("item %q not found: %w", itemName, err)
	}
	var item BitwardenItem
	if err := json.Unmarshal([]byte(out), &item); err != nil {
		return nil, fmt.Errorf("parsing item JSON: %w", err)
	}
	return &item, nil
}

// GetPassword retrieves the password of a login item by name.
func (c *BitwardenClient) GetPassword(itemName string) (string, error) {
	out, err := c.run("get", "password", itemName)
	if err != nil || out == "" {
		return "", fmt.Errorf("password for %q not found: %w", itemName, err)
	}
	return out, nil
}

// GetField retrieves a custom field from a named item.
func (c *BitwardenClient) GetField(itemName, fieldName string) (string, error) {
	item, err := c.GetItem(itemName)
	if err != nil {
		return "", err
	}
	for _, f := range item.Fields {
		if f.Name == fieldName {
			return f.Value, nil
		}
	}
	return "", fmt.Errorf("field %q not found in item %q", fieldName, itemName)
}

// ListItems returns all items matching an optional search term.
func (c *BitwardenClient) ListItems(search string) ([]BitwardenItem, error) {
	args := []string{"list", "items"}
	if search != "" {
		args = append(args, "--search", search)
	}
	out, err := c.run(args...)
	if err != nil {
		return nil, err
	}
	var items []BitwardenItem
	if err := json.Unmarshal([]byte(out), &items); err != nil {
		return nil, fmt.Errorf("parsing items JSON: %w", err)
	}
	return items, nil
}

func main() {
	client, err := NewBitwardenClient()
	if err != nil {
		log.Fatalf("Failed to create Bitwarden client: %v", err)
	}

	items, err := client.ListItems("github")
	if err != nil {
		log.Fatalf("ListItems error: %v", err)
	}
	fmt.Printf("Found %d item(s) matching 'github'\n", len(items))

	if len(items) > 0 {
		item := items[0]
		fmt.Printf("First match: %s\n", item.Name)

		password, err := client.GetPassword(item.Name)
		if err != nil {
			fmt.Printf("Could not retrieve password: %v\n", err)
		} else {
			fmt.Printf("Password retrieved (length: %d)\n", len(password))
		}
	}
}
